### DISCUSSION TITLE
setupCreateOperation - issue with translatable/tags

### DISCUSSION THREAD

**Discussion by alancwoo** [Laravel-Backpack/community-forum#691] [Q&A (Help)]
I am trying to use Backpack to manage [Spatie: laravel-tags](https://spatie.be/docs/laravel-tags/v4/introduction) via the [JSON field addon](https://github.com/ziming/json-field-for-backpack).

The tags store the `name` and `slug` as JSON blobs with locales like `{"en": "example"}`.

In `setupListOperation()`, I'm getting the plain string like so:

```
$this->crud->addColumn([
            'name' => 'name',
            'label' => 'name',
            'type' => 'text',
            'type' => 'model_function',
            'function_name' => 'getTranslation',
            'function_parameters' => ['name', 'en'],
        ]);
```

However, in the `setupCreateOperation()` the JSON field is broken because somehow the translation string is passed through not the JSON itself:

```
$this->crud->addField([
            'name' => 'name',
            'type' => 'json',
            'view_namespace' => 'json-field-for-backpack::fields',
            'modes' => ['code', 'form'],
        ]);
```

How can I ensure that the raw JSON is passed through here in the edit/create operation so it can be edited accordingly?

**Comment by pxpm** [comment]
Does a [model mutator]( https://laravel.com/docs/10.x/eloquent-mutators#defining-an-accessor) solve your problem here ?

**Reply by alancwoo** [reply]
Hm, actually now I realize it probably should function as a [Translatable model](https://backpackforlaravel.com/docs/6.x/crud-operation-update#translatable-models-and-multi-language-cruds) - and should thus be editable as a text field through locale/translation.

So I've tried to update my model accordingly:

```
...
use Backpack\CRUD\app\Models\Traits\SpatieTranslatable\HasTranslations;

class MyTag extends Tag
{
    use CrudTrait;
    use HasTranslations;
    ...
```

and my `TagCrudController` contains:
```
...
public function setup()
    {
        CRUD::setModel(\App\Models\MyTag::class);
...
```

but then encounter a number of errors regarding the inheritance/compatibility
> Declaration of Backpack\CRUD\app\Models\Traits\SpatieTranslatable\HasTranslations::getAttributeValue($key) must be compatible with Spatie\Tags\Tag::getAttributeValue($key): mixed

If I update `/Vendor/Backpack/crud/src/app/Models/Traits/SpatieTranslatable/HasTranslations.php` to match the return type hints:

```
public function getAttributeValue($key): mixed
    {
```

and keep correcting things I eventually encounter:

> Declaration of Backpack\CRUD\app\Models\Traits\SpatieTranslatable\HasTranslations::setLocale($locale) must be compatible with Spatie\Tags\Tag::setLocale(string $locale): Spatie\Tags\Tag

If I adjust this to read ` public function setLocale($locale): self`, backpack does finally load with translations:

```
        $this->crud->addField([
            'name' => 'name',
            'type' => 'text',
        ]);
        $this->crud->addField([
            'name' => 'slug',
            'type' => 'text',
        ]);
```
<img width="270" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/1481890/9656d0b5-c27a-4c4f-9c40-993ffefeb3e2">

but then upon saving, they are set to null:

<img width="342" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/1481890/778f3f9c-b9f5-4187-bba1-87a998ef060c">


Note: I've updated the title to reflect the issue more accurately

**Reply by alancwoo** [reply]
Hm, I've also set up a completely fresh laravel install with backpack and laravel-tags and the same thing occurs. However, creating a test model with translatable is working, so I am wondering if there is an issue with laravel-tags: https://github.com/spatie/laravel-tags/issues/482
